pallets/phala-world/src/nft_sale.rs

```diff
@@ -0,0 +1,1686 @@
+		);
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Magic,
+			RaceType::Cyborg,
+			magic_nft_sale_info.clone(),
+		);
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Magic,
+			RaceType::Pandroid,
+			magic_nft_sale_info.clone(),
+		);
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Magic,
+			RaceType::XGene,
+			magic_nft_sale_info,
+		);
+		let hero_nft_sale_info = NftSaleInfo {
+			race_count: 0,
+			race_for_sale_count: 1250,
+			race_giveaway_count: 50,
+			race_reserved_count: 0,
+		};
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Hero,
+			RaceType::AISpectre,
+			hero_nft_sale_info.clone(),
+		);
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Hero,
+			RaceType::Cyborg,
+			hero_nft_sale_info.clone(),
+		);
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Hero,
+			RaceType::Pandroid,
+			hero_nft_sale_info.clone(),
+		);
+		OriginOfShellsInventory::<T>::insert(
+			OriginOfShellType::Hero,
+			RaceType::XGene,
+			hero_nft_sale_info,
+		);
+		// Set IsOriginOfShellsInventorySet to true
+		IsOriginOfShellsInventorySet::<T>::put(true);
+
+		Ok(())
+	}
+
+	/// Update the NftSaleInfo for a given OriginOfShellType and RaceType
+	///
+	/// Parameters:
+	/// - `origin_of_shell_type`: OriginOfShellType to update in OriginOfShellInventory
+	/// - `race`: RaceType to update in OriginOfShellInventory
+	/// - `for_sale_count`: count to increment the for sale count
+	/// - `giveaway_count`: count to increment the race giveaway count
+	fn update_nft_sale_info(
+		origin_of_shell_type: OriginOfShellType,
+		race: RaceType,
+		for_sale_count: u32,
+		giveaway_count: u32,
+	) -> DispatchResult {
+		OriginOfShellsInventory::<T>::try_mutate_exists(
+			origin_of_shell_type,
+			race,
+			|nft_sale_info| -> DispatchResult {
+				if let Some(nft_sale_info) = nft_sale_info {
+					nft_sale_info.race_for_sale_count =
+						nft_sale_info.race_for_sale_count.saturating_add(for_sale_count);
+					nft_sale_info.race_giveaway_count =
+						nft_sale_info.race_giveaway_count.saturating_add(giveaway_count);
+				}
+				Ok(())
+			},
+		)?;
+
+		Ok(())
+	}
+
+	/// Set the origin of shell type, race and career attributes for a Origin of Shell NFT
+	///
+	/// Parameters:
+	/// - `collection_id`: Collection id of the Origin of Shell NFT
+	/// - `nft_id`: NFT id of the Origin of Shell NFT
+	/// - `origin_of_shell_type`: Origin of Shell type for the NFT
+	/// - `race`: Race attribute to set for the Origin of Shell NFT
+	/// - `career`: Career attribute to set for the Origin of Shell NFT
+	fn set_origin_of_shell_attributes(
+		collection_id: CollectionId,
+		nft_id: NftId,
+		origin_of_shell_type: OriginOfShellType,
+		race: RaceType,
+		career: CareerType,
+	) -> DispatchResult {
+		let overlord = Self::get_overlord_account()?;
+
+		let origin_of_shell_type_key: BoundedVec<u8, T::KeyLimit> =
+			self::Pallet::<T>::to_boundedvec_key("origin_of_shell_type")?;
```

Comment:
Does `self::to_boundedvec_key("origin_of_shell_type")?` work?

Reply:
This is the function:
```
fn to_boundedvec_key(name: &str) -> Result<BoundedVec<u8, T::KeyLimit>, Error<T>> {
	name.as_bytes().to_vec().try_into().map_err(|_| Error::<T>::KeyTooLong)
}
```
This sets the key so we can use `pallet_uniques` `set_attribute()` function.